core/opregion.c

```diff
@@ -38,571 +41,327 @@ void lai_write_opregion(lai_nsnode_t *field, lai_variable_t *source) {
     lai_panic("undefined field write: %s", lai_stringify_node_path(field));
 }
 
-void lai_read_field(lai_variable_t *destination, lai_nsnode_t *field) {
-    struct lai_instance *instance = lai_current_instance();
-    lai_nsnode_t *opregion = field->fld_region_node;
+// this assumes little endian
+static void lai_buffer_put_at(uint8_t *buffer, uint64_t value, size_t bit_offset, size_t num_bits) {
+    size_t in_byte_offset = bit_offset & 7;
+    size_t byte_offset = bit_offset >> 3;
+    size_t progress = 0;
+    while (progress < num_bits) {
+        uint8_t byte = (value >> progress) & 0xFF;
 
-    uint64_t offset, value, mask;
-    size_t bit_offset;
+        uint8_t low_byte = byte << in_byte_offset;
+        uint8_t hi_byte = byte >> (8 - in_byte_offset);
 
-    mask = ((uint64_t)1 << field->fld_size);
-    mask--;
-    offset = field->fld_offset / 8;
-    void *mmio;
+        buffer[byte_offset + progress / 8] |= low_byte;
+        buffer[byte_offset + 1 + progress / 8] |= hi_byte;
 
-    // these are for PCI
-    LAI_CLEANUP_VAR lai_variable_t bus_number = LAI_VAR_INITIALIZER;
-    LAI_CLEANUP_VAR lai_variable_t seg_number = LAI_VAR_INITIALIZER;
-    LAI_CLEANUP_VAR lai_variable_t address_number = LAI_VAR_INITIALIZER;
-    uint64_t bbn_result = 0; // When _BBN is not present, we assume PCI bus 0.
-    uint64_t seg_result = 0; // When _SEG is not present, we default to Segment Group 0
-    uint64_t adr_result = 0; // When _ADR is not present, again, default to zero.
-    size_t pci_byte_offset;
-
-    if (opregion->op_address_space != ACPI_OPREGION_PCI) {
-        switch (field->fld_flags & 0x0F) {
-            case FIELD_BYTE_ACCESS:
-                bit_offset = field->fld_offset % 8;
-                break;
-
-            case FIELD_WORD_ACCESS:
-                bit_offset = field->fld_offset % 16;
-                offset &= (~1);        // clear lowest bit
-                break;
-
-            case FIELD_DWORD_ACCESS:
-            case FIELD_ANY_ACCESS:
-                bit_offset = field->fld_offset % 32;
-                offset &= (~3);        // clear lowest two bits
-                break;
+        progress += 8;
+    }
+}
 
-            case FIELD_QWORD_ACCESS:
-                bit_offset = field->fld_offset % 64;
-                offset &= (~7);        // clear lowest three bits
-                break;
+// this assumes little endian
+static uint64_t lai_buffer_read_at(uint8_t *buffer, size_t byte_offset, size_t num_bytes) {
```

Comment:
It's confusing that this function deals with bytes while the `put_at` function deals with bits. Also, rename it to `get_at` for consistency.